src/Language/PureScript/BundleOpt.hs

```diff
@@ -0,0 +1,474 @@
+-----------------------------------------------------------------------------
+--
+-- Module      :  Language.PureScript.BundleOpt
+-- Copyright   :  (c) Jürgen Nicklisch-Franken 2016
+-- License     :  MIT
+--
+-- Maintainer  :  Jürgen Nicklisch <juergen.nicklisch@symbolian.net>
+-- Stability   :  experimental
+-- Portability :
+--
+-- | Optimize steps added to the bundle process for purescript.
+--
+-- This module takes as input the Javascript AST and applies optimizations like
+-- uncurrying and inlining
+-----------------------------------------------------------------------------
+
+{-# LANGUAGE PatternGuards #-}
+
+module Language.PureScript.BundleOpt (
+    uncurryFunc
+) where
+
+import Prelude.Compat
+import qualified Data.Map as M
+import Data.Maybe (mapMaybe)
+
+import Language.PureScript.BundleTypes
+import Language.JavaScript.Parser
+import Language.JavaScript.Parser.AST
+
+
+-- * Types
+
+data FuncAdmin = FuncAdmin
+    { moduleId  :: ModuleIdentifier
+    , arity     :: Int
+    , exported  :: Bool}
+    deriving (Eq,Show)
+
+type FuncAdminMap = M.Map String [FuncAdmin]
+
+data FuncStats = FuncStats {
+    moduleCount :: Int,
+    moduleFunctions :: Int,
+    uncurriedFuncs :: Int,
+    exportedEntities :: Int,
+    uncurriedFuncsExported :: Int,
+    exportedForeignEntities :: Int,
+    uncurriedForeignFuncsExported :: Int
+} deriving (Eq,Show)
+
+data FuncCollector = FuncCollector {
+    adminMap :: FuncAdminMap,
+    stats :: FuncStats
+} deriving (Eq,Show)
+
+-- * Constants
+
+suffix :: String
+suffix = "$_$_$"
+
+
+-- * Functions
+
+-- | Main function for uncurry optimization
+uncurryFunc :: [Module] -> [ModuleIdentifier] -> [Module]
+uncurryFunc modules entryPoints =
+    -- add uncurried functions
+    let (modulesWithUncurried,funcCollector) = foldr generateUncurried ([],emptyCollector) modules
```

Comment:
Yes, in principle, this could be expressed by a State monad. Since all function calls are cleanly encapsulated though, the State token (```funcCollector```) never leaves or enters any scope it's not supposed to be in, we chose not use a monad. After all, the transformations performed by these functions don't necessarily call for a monadic interface.